### PR TITLE
Add more asserts and docs to ChangeNotifier.

### PR DESCRIPTION
**Pull request by Hixie**
It took me a while to figure out what was going on (I was removing a
listener after disposal). These asserts helped.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -32,29 +32,44 @@ abstract class Listenable {
...
+/// notifications (where N is the number of listeners).
 class ChangeNotifier extends Listenable {
   List<VoidCallback> _listeners;
 
   /// Register a closure to be called when the object changes.
+  ///
+  /// This method must not be called after [dispose] has been called.
   @override
   void addListener(VoidCallback listener) {
+    assert(_listeners != const <VoidCallback>[]);
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: I probably would have made this condition into a `_debugFoo` getter.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/foundation/change_notifier.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: done, PTAL

**Comment by abarth** [comment]
LGTM

**Pull request merged**